### ISSUE TITLE
does inferencing on cpu use multi-threading ?

### ISSUE THREAD

**Issue by yassine017** [EndlessReform/fish-speech.rs#32]
can we somehow make the inferencing on cpu multi-threading ? it works fine , it just seems to only uses one core , I don't know if this should be an "enhancement" ,,, my rust has gotten a little rusty but I'll fiddle with it if I can

**Comment by EndlessReform** [comment]
It's a good question: it depends on what CPU type you're using. Under the hood this uses [candle.rs](https://github.com/huggingface/candle). However upstream, there are some reports for the default CPU target going inadvertently single-threaded for heavy computation in `gemm` (the matrix multiplies for the actual model) similar to what you're mentioning:
- https://github.com/huggingface/candle/issues/1103: Matrix multiplication dependencies can be inadvertently single-threaded
- https://github.com/huggingface/candle/issues/2571: someone else running into this

This is probably lower-level than I can realistically fix, sorry.

However, Candle has hardware acceleration for Apple CPUs with the `accelerate` flag and Intel with the `mkl` flag, I just need to test it and pass those flags through. It's possible this can be enabled and hopefully make CPU faster. Would you mind telling me which CPU hardware you're using?

**Comment by yassine017** [comment]
Oh , What a lovely team , Quick and snappy ! 
I admire the hard work ! 
yes , I'm running debian on an old 4th gen Intel 
CPU: Intel i5-4590 (4) @ 3.700GHz  , no the greatest , but my 14Gb of Memory handeled fish-speech 1.5 pretty well , and it seems to clone and work nominally , better than I anticipated , just noticed the single threaded-iness when it's going through generating audio streams.
I think you're right , those flags at compile time could lead to some CPUs to benifit from Multi-threading... not sure about this old fella tho!

**Comment by EndlessReform** [comment]
Sorry, I'm not actually able to replicate the single-threading issue. The only comparable hardware I have is Intel 7th gen (an old i5-7600k). It's pretty slow (only 5 tokens/sec), but still looks multi-threaded:

<img width="1141" alt="image" src="https://github.com/user-attachments/assets/a6c7b2bc-8916-4f9a-a08f-5d974e8f4518" />

```
Split into 2 chunks with progressive sizing
Chunks:
[TextChunk { text: "Hello world." }, TextChunk { text: "It's me, testing threading. can you confirm what's going on?" }]
Beginning chunk 0 of 2
1578.37ms prompt processing: 354 tokens (354 new, 0 cached, 224.28 tokens/s)
- Tokens: 20 [00:00:03] 5.1263/s iterations/s                                                                                                  21 tokens generated in 3.902s (5.38 tokens/s, 195.098ms / token, RTF: 0.250)
Vocoding took: 963 ms
Beginning chunk 1 of 2
341.88ms prompt processing: 367 tokens (25 new, 342 cached, 73.12 tokens/s)
\ Tokens: 80 [00:00:15] 5.0833/s iterations/s                                                                                                  81 tokens generated in 15.761s (5.14 tokens/s, 197.016ms / token, RTF: 0.239)
Vocoding took: 5698 ms
Generation complete
Final cache cleared
```

I'll still look into how difficult mkl support would be though

**Comment by yassine017** [comment]
![image](https://github.com/user-attachments/assets/ccacaf8e-f8bc-4745-a1a9-e35d7023966c)
mine is slower , but not by much , the server is indeed using up as many threads as it can , but the Generating process seems to hold up to only one , making it only able to use up 25% max of the CPU , It wouldn't be too noticeable on more modern hardware since candle might even allow full multi-threading... even on your 7th gen intel , I noticed that , while the server is indeed making use of a few threads , the process that generates seems to take up one thread ,,, I hope this isn't related to chunking and we're barking at the wrong tree

**Comment by EndlessReform** [comment]
No, it's definitely not the chunking slowing it down, speed on a 4090 is ~140-150 tokens/sec. Can you try the `mkl` branch I just pushed? It's about a 20-30% speedup on my machine (4 cores, 4 threads); CPU utilization seems to be full now:

<img width="1048" alt="image" src="https://github.com/user-attachments/assets/388495a2-8446-47bd-b7f0-bd1385eca975" />

What I had to do to get this working on Ubuntu 22.04 (copy-pasted from https://github.com/huggingface/candle/issues/443#issuecomment-1677533016):

```bash
wget -O- https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB \
| gpg --dearmor | sudo tee /usr/share/keyrings/oneapi-archive-keyring.gpg > /dev/null && \
  echo "deb [signed-by=/usr/share/keyrings/oneapi-archive-keyring.gpg] https://apt.repos.intel.com/oneapi all main" | \
  sudo tee /etc/apt/sources.list.d/oneAPI.list

sudo apt-get update
sudo apt-get install -y intel-oneapi-mkl-devel libomp-dev

git checkout mkl
cargo clean
# rest of args
cargo run --release --bin server --features mkl
```

```
[TextChunk { text: "Hello world." }, TextChunk { text: "It's me, testing threading. can you confirm what's going on? please help me i am stuck in this box" }]
Beginning chunk 0 of 2
1319.81ms prompt processing: 354 tokens (354 new, 0 cached, 268.22 tokens/s)
/ Tokens: 21 [00:00:03] 6.8205/s iterations/s                                                                                                   22 tokens generated in 3.124s (7.04 tokens/s, 148.766ms / token, RTF: 0.327)
Vocoding took: 781 ms
Beginning chunk 1 of 2
262.07ms prompt processing: 376 tokens (34 new, 342 cached, 129.73 tokens/s)
| Tokens: 128 [00:00:20] 6.3582/s iterations/s                                                                                                  129 tokens generated in 20.123s (6.41 tokens/s, 157.214ms / token, RTF: 0.298)
Vocoding took: 8808 ms
Generation complete
Final cache cleared
```

**Comment by EndlessReform** [comment]
Note that looking at upstream, this seems to be a problem with the framework, people are reporting very similar problems with BERT and Llama. Afraid I probably can't do much more than this

**Comment by yassine017** [comment]
true ! all the way to gemm ,,, I'll try my best when I can , or just slap a GPU on it , I don't see a reason for this issue to remain open for too long , thanks for the much appreciated help Jacob @EndlessReform

**Comment by EndlessReform** [comment]
Thanks a lot for testing this out! Hope you enjoyed the codebase. Sorry this is slow on CPU - the official code with torch.compile might be a lot more faster. Just one more thing before I close the issue though, did the MKL flag at least help a little?
